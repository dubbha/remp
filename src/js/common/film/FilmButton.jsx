import React from 'react';
import PropTypes from 'prop-types';

const SearchButton = ({ onSearchClick }) => (
  <button
    className="film__button"
    onClick={onSearchClick}
  >
    Search
  </button>
);

SearchButton.propTypes = {
  onSearchClick: PropTypes.func.isRequired,
};

export default SearchButton;
