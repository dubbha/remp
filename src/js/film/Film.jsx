import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import FilmHeader from '../common/filmHeader';
import Result from '../common/result';
import List from '../common/list';
import Footer from '../common/footer';
import filmPropShape from '../common/utils/propShapes';
import * as selectors from '../common/store/selectors';
import * as actions from '../common/store/actions';

export class Film extends Component {
  static defaultProps = {
    film: null,
    filteredResults: null,
  };

  static propTypes = {
    match: PropTypes.shape({
      params: PropTypes.shape({
        id: PropTypes.string,
        title: PropTypes.string,
      }),
    }).isRequired,
    history: PropTypes.shape({
      push: PropTypes.func,
    }).isRequired,
    film: PropTypes.shape(filmPropShape),
    filteredResults: PropTypes.arrayOf(PropTypes.shape(filmPropShape)),
    getFilm: PropTypes.func.isRequired,
    getFilmDetails: PropTypes.func.isRequired,
    searchByDirector: PropTypes.func.isRequired,
    setSearchBy: PropTypes.func.isRequired,
    isLoading: PropTypes.bool.isRequired,
    isFilmLoading: PropTypes.bool.isRequired,
  };

  static fetchData = (dispatch, match) =>
    dispatch(this.props.getFilm(match.params.id));

  componentWillMount = () => {
    const {
      match: { params },
      film,
      filteredResults,
      getFilm,
      getFilmDetails,
      searchByDirector,
    } = this.props;

    const id = +params.id;

    if (!film) {
      getFilm(id);
    } else {
      if (!film.runtime || !film.cast || !film.director) {
        getFilmDetails(film);
      }
      if (filteredResults && filteredResults.length === 0 && film.director) {
        searchByDirector(film.director);
      }
    }
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.match.params.id !== this.props.match.params.id) {
      const film = nextProps.film;
      const id = nextProps.match.params.id;
      const { getFilm, getFilmDetails } = this.props;

      if (!film) {
        getFilm(id);
      } else if (!film.runtime || !film.cast || !film.director) {
        getFilmDetails(film);
      }
    }
  }

  handleSelectFilm = (film) => {
    const { history } = this.props;

    window.scrollTo(0, 0);
    history.push(`/film/${film.id}/${encodeURIComponent(film.title)}`);
  }

  handleSearchClick = () => {
    const { history, setSearchBy, film } = this.props;

    window.scrollTo(0, 0);

    if (film.director) {
      setSearchBy('director');
      history.push(`/search/${encodeURIComponent(film.director)}`);
    } else {
      history.push('/search');
    }
  }

  render() {
    const { isLoading, isFilmLoading, film, filteredResults } = this.props;

    return (
      <div className="app__container">
        {
          film && (
            <FilmHeader
              film={film}
              onSearchClick={this.handleSearchClick}
            />
          )
        }
        {
          film && !isFilmLoading && (
            <Result
              film={film}
              results={filteredResults}
            />
          )
        }
        <List
          results={filteredResults}
          onSelectFilm={this.handleSelectFilm}
          isLoading={isLoading || isFilmLoading}
        />
        <Footer />
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  film: selectors.filmSelector(state, props),
  filteredResults: selectors.filteredResultsSelector(state, props),
  isLoading: selectors.isLoadingSelector(state),
  isFilmLoading: selectors.isFilmLoadingSelector(state),
});

const mapDispatchToProps = {
  getFilm: actions.getFilm,
  getFilmDetails: actions.getFilmDetails,
  searchByDirector: actions.searchByDirector,
  setSearchBy: actions.setSearchBy,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Film);
