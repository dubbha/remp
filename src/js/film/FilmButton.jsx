import React from 'react';
import PropTypes from 'prop-types';

const SearchButton = ({ onClearFilm }) => (
  <button
    className="film__button"
    onClick={onClearFilm}
  >
    Search
  </button>
);

SearchButton.propTypes = {
  onClearFilm: PropTypes.func.isRequired,
};

export default SearchButton;
