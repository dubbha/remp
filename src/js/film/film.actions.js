import axios from 'axios';
import { apiUrl, apiKey } from '../common/config/api.config';
import * as searchActions from '../search/search.actions';

export const actionTypes = {
  SET_FILM: 'SET_FILM',
  SET_IS_LOADING: 'SET_IS_LOADING',
};

export const setFilm = film => ({
  type: actionTypes.SET_FILM,
  film,
});

export const setIsLoading = isLoading => ({
  type: actionTypes.SET_IS_LOADING,
  isLoading,
});

export const getFilm = query => (dispatch) => {
  dispatch(setIsLoading(true));

  return axios.get(`${apiUrl}search/movie`,
    {
      params: {
        query,
        api_key: apiKey,
      },
    })
    .then((res) => {
      if (res.data && res.data.results && res.data.results.length) {
        const film = res.data.results[0];
        axios.get(`${apiUrl}movie/${film.id}`,
          {
            params: {
              query,
              api_key: apiKey,
              append_to_response: 'credits',
            },
          })
          .then((res2) => {
            if (res2.data && res2.data.credits && res2.data.credits.crew) {
              const directorArr = res2.data.credits.crew.filter(i => i.job === 'Director');

              if (directorArr.length) { // some movies have no director
                film.director = directorArr[0].name;
              }

              film.runtime = res2.data.runtime;
              film.cast = res2.data.credits.cast;

              dispatch(setFilm(film));
              dispatch(setIsLoading(false));

              if (film.director) {
                dispatch(searchActions.searchByDirector(film.director));
              }

              dispatch(searchActions.setResultDetails(film.id, { film }));
            }
          });
      }
    });
};

export const getFilmDetails = film => dispatch =>
  axios.get(`${apiUrl}movie/${film.id}`,
    {
      params: {
        api_key: apiKey,
        append_to_response: 'credits',
      },
    })
    .then((res) => {
      if (res.data && res.data) {
        const runtime = res.data.runtime;
        const cast = res.data.credits.cast;

        if (!film.director) {
          const directorArr = res.data.credits.crew.filter(i => i.job === 'Director');

          if (directorArr.length) { // some movies have no director
            const director = directorArr[0].name;

            dispatch(setFilm({ ...film, runtime, cast, director }));

            dispatch(searchActions.searchByDirector(director));
            dispatch(searchActions.setResultDetails(film.id, { runtime, cast, director }));
          }
        } else {
          dispatch(setFilm({ ...film, runtime, cast }));
          dispatch(searchActions.setResultDetails(film.id, { runtime, cast }));
        }
      }
    });

export const selectFilm = film => (dispatch) => {
  dispatch(setFilm(film)); // display available data

  if (!film.runtime || !film.cast || !film.director) {
    dispatch(getFilmDetails(film)); // fetch additional data
  }
};
