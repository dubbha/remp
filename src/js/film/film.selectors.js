import { createSelector } from 'reselect';
import { selectors as searchSelectors } from '../search';

const subStateSelector = state => state.film;

export const filmSelector = createSelector(
  subStateSelector,
  ({ film }) => film,
);

export const filteredResultsSelector = createSelector(
  [searchSelectors.resultsSelector, filmSelector],
  (results, film) => results.filter(item => item.show_title !== film.title),
);
