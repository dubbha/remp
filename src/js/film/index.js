import reducer from '../common/store/reducer';
import * as actions from '../common/store/actions';
import * as selectors from '../common/store/selectors';

export { default } from './Film';

export {
  reducer,
  actions,
  selectors,
};
