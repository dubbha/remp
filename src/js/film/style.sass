.film
    background-image: url('../../img/bg.jpg')
    background-size: cover
    min-height: 450px
