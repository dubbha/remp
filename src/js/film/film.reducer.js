import { actionTypes } from './film.actions';

const initialState = {
  film: null,
  isLoading: false,
};

const filmReducer = (state = initialState, action) => {
  switch (action.type) {
    case actionTypes.SET_FILM:
      return {
        ...state,
        film: action.film,
      };
    case actionTypes.SET_IS_LOADING:
      return {
        ...state,
        isLoading: action.isLoading,
      };
    default:
      return state;
  }
};

export default filmReducer;
