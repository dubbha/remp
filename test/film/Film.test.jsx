import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { connect } from 'react-redux';
import { Film } from 'film/Film';

jest.mock('react-redux', () => ({
  connect: jest.fn(() => jest.fn()),
}));

jest.mock('common/store/selectors', () => ({
  filmSelector: jest.fn((state, props) => state.results.find(i => i.id === +props.match.params.id)),
  filteredResultsSelector: jest.fn((state, props) => state.results.filter(i => i.id !== +props.match.params.id)),
  isLoadingSelector: jest.fn(state => state.isLoading),
  isFilmLoadingSelector: jest.fn(state => state.isFilmLoading),
}));

jest.mock('common/store/actions', () => ({
  getFilm: jest.fn(id => `film with id ${id}`),
}));

describe('Film', () => {
  describe('react-redux connector', () => {
    it('should map state to props', () => {
      const mapStateToProps = connect.mock.calls[0][0];

      const state = {
        isLoading: false,
        isFilmLoading: false,
        results: [
          { id: 1, title: 'TITLE1' },
          { id: 2, title: 'TITLE2' },
          { id: 3, title: 'TITLE3' },
        ],
      };

      const props = {
        match: {
          params: { id: '1' },
        },
      };

      expect(mapStateToProps(state, props)).toEqual({
        film: { id: 1, title: 'TITLE1' },
        filteredResults: [
          { id: 2, title: 'TITLE2' },
          { id: 3, title: 'TITLE3' },
        ],
        isLoading: false,
        isFilmLoading: false,
      });
    });
  });

  describe('unconnected component', () => {
    let props;
    beforeEach(() => {
      props = {
        match: { params: { id: '1', title: 'TITLE' } },
        history: { push: jest.fn() },
        film: {
          id: 1,
          title: 'TITLE',
          runtime: 120,
          director: 'DIRECTOR',
          cast: [
            { character: 'CHARACTER1', name: 'NAME1' },
            { character: 'CHARACTER2', name: 'NAME2' },
          ],
        },
        filteredResults: [
          { id: 2, title: 'TITLE2' },
          { id: 3, title: 'TITLE3' },
        ],
        getFilm: jest.fn(),
        getFilmDetails: jest.fn(),
        searchByDirector: jest.fn(),
        setSearchBy: jest.fn(),
        isLoading: false,
        isFilmLoading: false,
      };
    });

    it('should render successfully', () => {
      const wrapper = shallow(<Film {...props} />);
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should not render result block if film is loading', () => {
      const wrapper = shallow(<Film {...props} isFilmLoading />);
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should not render film header and result blocks if film is not available', () => {
      const wrapper = shallow(<Film {...props} film={null} />);
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should get film if film is not available on mount', () => {
      props.film = null;
      const instance = new Film(props);

      instance.componentWillMount();

      expect(props.getFilm).toBeCalled();
    });

    it('should get film details if runtime is not available on mount', () => {
      props.film.runtime = null;
      const instance = new Film(props);

      instance.componentWillMount();

      expect(props.getFilmDetails).toBeCalled();
    });

    it('should get film details if cast runtime is not available on mount', () => {
      props.film.cast = null;
      const instance = new Film(props);

      instance.componentWillMount();

      expect(props.getFilmDetails).toBeCalled();
    });

    it('should get film details if director is not available on mount', () => {
      props.film.director = null;
      const instance = new Film(props);

      instance.componentWillMount();

      expect(props.getFilmDetails).toBeCalled();
    });

    it('should search by director if film director is available but results are empty on mount', () => {
      props.film = { id: 1, title: 'TITLE', director: 'DIRECTOR' };
      props.filteredResults = [];
      const instance = new Film(props);

      instance.componentWillMount();

      expect(props.searchByDirector).toBeCalled();
    });

    it('should get film if route id param changed but film is not available on props change', () => {
      const instance = new Film(props);

      instance.componentWillReceiveProps({
        match: { params: { id: '2', title: 'TITLE2' } },
        film: null,
      });

      expect(props.getFilm).toBeCalled();
    });

    it('should get film details if route id param changed but film runtime is not available on props change', () => {
      const instance = new Film(props);

      instance.componentWillReceiveProps({
        match: { params: { id: '2', title: 'TITLE2' } },
        film: {
          id: 2,
          title: 'TITLE2',
          director: 'DIRECTOR',
          cast: [
            { character: 'CHARACTER1', name: 'NAME1' },
            { character: 'CHARACTER2', name: 'NAME2' },
          ],
        },
      });

      expect(props.getFilmDetails).toBeCalled();
    });

    it('should get film details if route id param changed but film cast is not available on props change', () => {
      const instance = new Film(props);

      instance.componentWillReceiveProps({
        match: { params: { id: '2', title: 'TITLE2' } },
        film: {
          id: 2,
          title: 'TITLE2',
          runtime: 120,
          director: 'DIRECTOR',
        },
      });

      expect(props.getFilmDetails).toBeCalled();
    });

    it('should get film details if route id param changed but film director is not available on props change', () => {
      const instance = new Film(props);

      instance.componentWillReceiveProps({
        match: { params: { id: '2', title: 'TITLE2' } },
        film: {
          id: 2,
          title: 'TITLE2',
          runtime: 120,
          cast: [
            { character: 'CHARACTER1', name: 'NAME1' },
            { character: 'CHARACTER2', name: 'NAME2' },
          ],
        },
      });

      expect(props.getFilmDetails).toBeCalled();
    });

    it('should not get film if route id param has not changed on props change', () => {
      const instance = new Film(props);

      instance.componentWillReceiveProps({
        match: { params: { id: '1', title: 'TITLE' } },
        film: null,
      });

      expect(props.getFilm).not.toBeCalled();
    });

    it('should not get film details if route id param has not changed on props change', () => {
      const instance = new Film(props);

      instance.componentWillReceiveProps({
        match: { params: { id: '1', title: 'TITLE' } },
        film: {
          id: 1,
          title: 'TITLE',
        },
      });

      expect(props.getFilmDetails).not.toBeCalled();
    });

    it('should not get film details if route id param changed but film details are available on props change', () => {
      const instance = new Film(props);

      instance.componentWillReceiveProps({
        match: { params: { id: '2', title: 'TITLE2' } },
        film: {
          id: 2,
          title: 'TITLE2',
          runtime: 120,
          cast: [
            { character: 'CHARACTER1', name: 'NAME1' },
            { character: 'CHARACTER2', name: 'NAME2' },
          ],
          director: 'DIRECTOR',
        },
      });

      expect(props.getFilmDetails).not.toBeCalled();
    });

    it('should handle film selection', () => {
      jest.spyOn(global.window, 'scrollTo');

      const wrapper = shallow(<Film {...props} />);

      wrapper.find('List').simulate('selectFilm', props.film);

      expect(global.window.scrollTo).toBeCalledWith(0, 0);
    });

    it('should push to route history on handle film selection', () => {
      const wrapper = shallow(<Film {...props} />);

      wrapper.find('List').simulate('selectFilm', props.film);

      expect(props.history.push).toBeCalledWith('/film/1/TITLE');
    });

    it('should handle search click', () => {
      jest.spyOn(global.window, 'scrollTo');

      const wrapper = shallow(<Film {...props} />);

      wrapper.find('FilmHeader').simulate('searchClick');

      expect(global.window.scrollTo).toBeCalledWith(0, 0);
    });

    it('should set searchBy to director if film director is available on search click', () => {
      const wrapper = shallow(<Film {...props} />);

      wrapper.find('FilmHeader').simulate('searchClick');

      expect(props.setSearchBy).toBeCalledWith('director');
    });

    it('should push to route history with director name if available on handle search click', () => {
      const wrapper = shallow(<Film {...props} />);

      wrapper.find('FilmHeader').simulate('searchClick');

      expect(props.history.push).toBeCalledWith('/search/DIRECTOR');
    });

    it('should push to route history if director is not available on handle search click', () => {
      props.film.director = null;
      const wrapper = shallow(<Film {...props} />);

      wrapper.find('FilmHeader').simulate('searchClick');

      expect(props.history.push).toBeCalledWith('/search');
    });
  });

  describe('fetchData static method', () => {
    it('should dispatch an action required for server-side rendering data pre-fetch', () => {
      const match = { params: { id: '1' } };
      const dispatch = jest.fn();

      Film.fetchData(dispatch, match);

      expect(dispatch).toBeCalledWith('film with id 1');
    });
  });
});
