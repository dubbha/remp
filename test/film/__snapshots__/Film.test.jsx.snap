// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Film unconnected component should not render film header and result blocks if film is not available 1`] = `
<div
  className="app__container"
>
  <List
    film={Object {}}
    isLoading={false}
    onSelectFilm={[Function]}
    results={
      Array [
        Object {
          "id": 2,
          "title": "TITLE2",
        },
        Object {
          "id": 3,
          "title": "TITLE3",
        },
      ]
    }
  />
  <Footer />
</div>
`;

exports[`Film unconnected component should not render result block if film is loading 1`] = `
<div
  className="app__container"
>
  <FilmHeader
    film={
      Object {
        "cast": Array [
          Object {
            "character": "CHARACTER1",
            "name": "NAME1",
          },
          Object {
            "character": "CHARACTER2",
            "name": "NAME2",
          },
        ],
        "director": "DIRECTOR",
        "id": 1,
        "runtime": 120,
        "title": "TITLE",
      }
    }
    onSearchClick={[Function]}
  />
  <List
    film={Object {}}
    isLoading={true}
    onSelectFilm={[Function]}
    results={
      Array [
        Object {
          "id": 2,
          "title": "TITLE2",
        },
        Object {
          "id": 3,
          "title": "TITLE3",
        },
      ]
    }
  />
  <Footer />
</div>
`;

exports[`Film unconnected component should render successfully 1`] = `
<div
  className="app__container"
>
  <FilmHeader
    film={
      Object {
        "cast": Array [
          Object {
            "character": "CHARACTER1",
            "name": "NAME1",
          },
          Object {
            "character": "CHARACTER2",
            "name": "NAME2",
          },
        ],
        "director": "DIRECTOR",
        "id": 1,
        "runtime": 120,
        "title": "TITLE",
      }
    }
    onSearchClick={[Function]}
  />
  <Result
    film={
      Object {
        "cast": Array [
          Object {
            "character": "CHARACTER1",
            "name": "NAME1",
          },
          Object {
            "character": "CHARACTER2",
            "name": "NAME2",
          },
        ],
        "director": "DIRECTOR",
        "id": 1,
        "runtime": 120,
        "title": "TITLE",
      }
    }
    onSortByChange={null}
    results={
      Array [
        Object {
          "id": 2,
          "title": "TITLE2",
        },
        Object {
          "id": 3,
          "title": "TITLE3",
        },
      ]
    }
    sortBy={null}
    sortByParams={null}
  />
  <List
    film={Object {}}
    isLoading={false}
    onSelectFilm={[Function]}
    results={
      Array [
        Object {
          "id": 2,
          "title": "TITLE2",
        },
        Object {
          "id": 3,
          "title": "TITLE3",
        },
      ]
    }
  />
  <Footer />
</div>
`;
